File: src/main/native/src/io/writables/nat/heapitem.c
#include "heapitem.h"
#include "alloc.h"

hpi *hpiNew(void *data, hpi *neighbor){
    hpi *i = mclAlloc(sizeof(hpi));

    i->data = data;
    i->child = NULL;
    i->parent = NULL;
    i->degree = 0;

    if(neighbor){
        hpiInsertSibling(neighbor, i);
    } else {
        i->left = i;
        i->right =  i;
    }

    return i;
}

void hpiInsertSibling(hpi* root, hpi* item){
    item->right = root;
    item->left = root->left;
    root->left = item;
    item->left->right = item;
}

void hpiLink(hpi *parent, hpi *node){
    node->left->right = node->right;
    node->right->left = node->left;
    node->parent = parent;

    if(parent->child == NULL){
        parent->child = node;
        node->right = node;
        node->left = node;
    } else {
        node->left = parent->child;
        node->right = parent->child->right;
        parent->child->right = node;
        node->right->left = node;
    }

    parent->degree++;
}

void hpiFree(hpi **node){
    hpi *child = (*node)->child;
    if(child){
        child->left->right = NULL;

        for(hpi *i = child, *n; i; i = n){
            n = i->right;
            hpiFree(&i);
        }
    }

    mclFree(*node);
    *node = NULL;
}

File: src/main/java/io/heap/HeapItem.java
package io.heap;

public final class HeapItem {

	private final Object data;
	private HeapItem child = null;
	private HeapItem left;
	private HeapItem right;
	private int degree = 0;
	
	public HeapItem(Comparable<?> data, HeapItem nb) {
		this.data = data;
		
		if(nb == null){
			left = this;
			right = this;
		} else {
			right = nb;
			left = nb.left;
			nb.left = this;
			left.right = this;				
		}
	}
	
	public int degree() {
		return degree;
	}

	/**
	 * remove node from its siblings
	 */
	public void decouple(){
		left.right = right;
		right.left = left;
	}
	
	public void link(HeapItem parent){
		left.right = right;
		right.left = left;
		
		if(parent.child == null){
			parent.child = this;
			right = this;
			left = this;
		} else {
			left = parent.child;
			right = parent.child.right;
			parent.child.right = this;
			right.left = this;				
		}
	}
	
	public boolean hasChild(){
		return child != null;
	}
	
	public HeapItem child(){
		return child;
	}
	
	public HeapItem right(){
		return right;
	}
	
	public void takeDownChildren(){
		HeapItem min_left = left;
		HeapItem z_child_left = child.left;
		left = z_child_left;
		z_child_left.right = this;
		child.left = min_left;
		min_left.right = child;
	}
	
	public Object data(){
		return data;
	}
	
	public boolean hasSiblings(){
		return this != right;
	}
}